Implement Docker healthcheck script in Go

nitisht opened the pull request:
## Description
Implement Docker healthcheck script in Go

## Motivation and Context
Go script makes it easy to read/maintain. Also updated the timeout
in Dockerfiles from 5s to default 30s.

Higher timeout makes sense as server may sometimes respond slowly
if under high load as reported in #6974

Fixes #6974

## Regression
No

## How Has This Been Tested?
Local Docker test
- Build local Docker image using `TAG=custom-minio make docker`
- Run the image and confirm healthcheck status using `docker ps`

Swarm testing
- Deploy a Swarm cluster on https://labs.play-with-docker.com
- Confirm Minio cluster is deployed with healthy status

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

harshavardhana commented on Dockerfile:
```diff
@@ -11,7 +11,9 @@ RUN  \
-     go install -v -ldflags "$(go run buildscripts/gen-ldflags.go)"
+     go install -v -ldflags "$(go run buildscripts/gen-ldflags.go)" && \
+     cd /go/src/github.com/minio/minio/dockerscripts && \
+     go build -ldflags "-s-w" -o /usr/bin/healthcheck healthcheck.go 
```
looks like `-s-w` should be `-s -w` - missing a space?

harshavardhana commented on Dockerfile:
```diff
@@ -11,7 +11,9 @@ RUN  \
      cd /go/src/github.com/minio/minio && \
-     go install -v -ldflags "$(go run buildscripts/gen-ldflags.go)"
+     go install -v -ldflags "$(go run buildscripts/gen-ldflags.go)" && \
+     cd /go/src/github.com/minio/minio/dockerscripts && \
```
No need to `cd` you can simply build `go build dockerscripts/healthcheck.go`

harshavardhana commented on Dockerfile:
```diff
@@ -32,7 +34,6 @@ ENTRYPOINT ["/usr/bin/docker-entrypoint.sh"]
 
-HEALTHCHECK --interval=30s --timeout=5s \
-    CMD /usr/bin/healthcheck.sh
+HEALTHCHECK --interval=2m CMD healthcheck
```
why 2m?

nitisht replied:
with improvements in liveness handler we can probably reduce it, but I think checking the server for liveness every 30s is too frequent. Changing to 1m

harshavardhana commented on Dockerfile:
```diff
@@ -22,7 +24,7 @@ ENV MINIO_ACCESS_KEY_FILE=access_key \
 EXPOSE 9000
 
 COPY --from=0 /go/bin/minio /usr/bin/minio
```
are you sure healthcheck is getting copied in multi-stage build?

nitisht replied:
`COPY --from=0` was missing, fixed now

harshavardhana commented on Dockerfile.release:
```diff
@@ -1,8 +1,21 @@
 LABEL maintainer="Minio Inc <dev@minio.io>"
 
-COPY dockerscripts/docker-entrypoint.sh dockerscripts/healthcheck.sh /usr/bin/
+COPY dockerscripts/docker-entrypoint.sh /usr/bin/
```
are you sure healthcheck is getting copied in multi-stage build?

nitisht replied:
fixed.

nitisht committed:
Implement Docker healthcheck script in Go

Go script makes it easy to read/maintain. Also updated the timeout
in Dockerfiles from 5s to default 30s and test interval to 2m

Higher timeout makes sense as server may sometimes respond slowly
if under high load as reported in #6974

Fixes #6974

harshavardhana reviewed (dismissed):
LGTM

vadmeste commented on dockerscripts/healthcheck.go:
```diff
@@ -0,0 +1,122 @@
+			// exit with success
+			os.Exit(0)
+		}
+		bodyBytes, err := ioutil.ReadAll(resp.Body)
```
Error checking here is missing

nitisht replied:
fixed

harshavardhana replied:
We need to limit to a predefined size such that we don't consume a lot of memory here. Perhaps using a is.LimitReader is better.

vadmeste commented on dockerscripts/healthcheck.go:
```diff
@@ -0,0 +1,122 @@
+				// GET failed exit
+				os.Exit(1)
+			}
+			if resp.StatusCode == http.StatusOK {
```
missing os.Exit(1) for `else {}`

nitisht replied:
done

vadmeste commented on dockerscripts/healthcheck.go:
```diff
@@ -0,0 +1,122 @@
+		if resp.StatusCode == http.StatusForbidden && bodyString == "SSL required" {
+			// Try with https scheme
+			uri = strings.Join([]string{schemeSecure, endPoint, healthPath}, "")
+			resp, err = http.Get(uri)
```
We probably need to find a way to make this works in case of self signed certificate (although a workaround for the user is to install the certificate in the OS)

vadmeste commented on dockerscripts/healthcheck.go:
```diff
@@ -0,0 +1,122 @@
+		if resp.StatusCode == http.StatusForbidden && bodyString == "SSL required" {
+			// Try with https scheme
+			uri = strings.Join([]string{schemeSecure, endPoint, healthPath}, "")
+			resp, err = http.Get(uri)
```
resp.Body is not closed properly.

nitisht replied:
done..

vadmeste reviewed (commented):
Some comments

harshavardhana commented on dockerscripts/healthcheck.go:
```diff
@@ -0,0 +1,144 @@
+	} else {
+		endPoint := findEndpoint()
+		uri := strings.Join([]string{scheme, endPoint, healthPath}, "")
+		resp, err := http.Get(uri)
```
By default http.Get has no timeouts, we should timeout here otherwise there is a potential for hang. Instantiate a new http.Client with some dial timeouts etc.

nitisht replied:
we do have the timeout in Dockerfile itself i.e. `HEALTHCHECK` by default times out after `30s`. Wouldn't that be enough?

harshavardhana commented on dockerscripts/healthcheck.go:
```diff
@@ -0,0 +1,144 @@
+		// This means sever is configured with https
+		if resp.StatusCode == http.StatusForbidden && bodyString == "SSL required" {
+			// Try with https scheme
+			uri = strings.Join([]string{schemeSecure, endPoint, healthPath}, "")
```
Strings join is not a good idea, we should construct using url.URL and then using u.String()

harshavardhana commented on dockerscripts/healthcheck.go:
```diff
@@ -0,0 +1,144 @@
+const (
+	initGraceTime = 300
+	healthPath    = "/minio/health/live"
+	scheme        = "http://"
```
If we use url.URL we can avoid `://`

harshavardhana commented on dockerscripts/healthcheck.go:
```diff
@@ -0,0 +1,144 @@
+		os.Exit(1)
+	}
+	// split netstat output in rows
+	lines := strings.Split(string(op), "\n")
```
You can instead use bufio.Scanner here to look at line by line combine this with the os.Stdin reader from exec.Command() is more idiomatic style.

harshavardhana commented on dockerscripts/healthcheck.go:
```diff
@@ -0,0 +1,160 @@
+	}
+	if err = cmd.Wait(); err != nil {
+		// command failed to run
+		os.Exit(1)
```
Do you want to fail with log.Fatal instead ?

harshavardhana commented on dockerscripts/healthcheck.go:
```diff
@@ -0,0 +1,160 @@
+			// Could not parse URL successfully
+			os.Exit(1)
+		}
+		resp, err := http.Get(u.String())
```
We need timeouts here for http client, Default client doesn't have sufficient timeouts.

harshavardhana commented on dockerscripts/healthcheck.go:
```diff
@@ -0,0 +1,160 @@
+			os.Exit(1)
+		}
+		bodyString := string(bodyBytes)
+		resp.Body.Close()
```
Drain the body to avoid leaking connections.. and proper reuse.

vadmeste commented on dockerscripts/healthcheck.go:
```diff
@@ -0,0 +1,169 @@
+	scanner.Split(bufio.ScanLines)
+	// loop over the rows to find Minio process
+	for scanner.Scan() {
+		if strings.Contains(scanner.Text(), "minio") {
```
Can you make 'minio' as constant in the beginning of this file ?

nitisht replied:
done

vadmeste reviewed (dismissed):
LGTM & tested

harshavardhana committed:
Merge branch 'master' into go-healthcheck

vadmeste reviewed (approved):
LGTM

nitisht committed:
Merge branch 'master' into go-healthcheck
